File: View/FXDAnnotationView.m


#import "FXDAnnotationView.h"


@implementation MKAnnotationView (Essential)
- (instancetype)initWithAnnotation:(id<MKAnnotation>)annotation reuseIdentifier:(NSString *)reuseIdentifier withDefaultImage:(UIImage*)defaultImage shouldChangeOffset:(BOOL)shouldChangeOffset {
	
	//MARK: Need to use self, since this is an added category
	self = [self initWithAnnotation:annotation reuseIdentifier:reuseIdentifier];
	
	if (self) {
		self.image = defaultImage;
		
		if (defaultImage && shouldChangeOffset) {
			CGPoint modifiedOffset = self.centerOffset;
			modifiedOffset.y -= (defaultImage.size.height/2.0);
			[self setCenterOffset:modifiedOffset];
		}
	}
	
	return self;
}

- (void)animateCustomDropAfterDelay:(NSTimeInterval)delay fromOffset:(CGPoint)offset {
	CGRect animatedFrame = self.frame;
	
	// Move annotation out of view	
	CGRect modifiedFrame = self.frame;
	modifiedFrame.origin.x += offset.x;
	modifiedFrame.origin.y += offset.y;
	[self setFrame:modifiedFrame];
	
	// Animate drop
	[UIView
	 animateWithDuration:durationAnimation
	 delay:delay
	 options:UIViewAnimationOptionCurveEaseOut
	 animations:^{
		 [self setFrame:animatedFrame];
	 }
	 completion:^(BOOL finished){
		 
		 [UIView
		  animateWithDuration:durationQuickAnimation
		  animations:^{
			  self.transform = CGAffineTransformMakeScale(1.0, 0.8);
		  }
		  completion:^(BOOL finished){
			  [UIView
			   animateWithDuration:durationQuickAnimation
			   animations:^{
				   self.transform = CGAffineTransformIdentity;
			   }];
		  }];
	 }];
}

@end

File: Sources/fXDKit/Classes/FXDAnnotationView.swift
import MapKit


extension MKAnnotationView {
	public convenience init(annotation: MKAnnotation!, reuseIdentifier: String!, withDefaultImage defaultImage: UIImage?, shouldChangeOffset: Bool) {
		self.init(annotation: annotation, reuseIdentifier: reuseIdentifier)

		self.image = defaultImage

		if defaultImage != nil && shouldChangeOffset == true {
			self.centerOffset.y -= ((defaultImage?.size.height ?? 0.0)/2.0)
		}
	}


	public func animateCustomDrop(afterDelay delay: TimeInterval, fromOffset offset: CGPoint) {
		let animatedFrame = self.frame

		// Move annotation out of view
		var modifiedFrame = self.frame
		modifiedFrame.origin.x += offset.x
		modifiedFrame.origin.y += offset.y
		self.frame = modifiedFrame

		// Animate drop
		UIView.animate(withDuration: DURATION_ANIMATION,
					   delay: delay,
					   options: .curveEaseOut) {
			self.frame = animatedFrame

		} completion: {
			finished in

			UIView.animate(withDuration: DURATION_QUICK_ANIMATION) {
				self.transform = CGAffineTransform(scaleX: 1.0, y: 0.8)

			} completion: {
				finished in

				UIView.animate(withDuration: DURATION_QUICK_ANIMATION) {
					self.transform = .identity
				}
			}
		}
	}
}
